src/consumer.js

```
@@ -0,0 +1,265 @@
+   * completed (ack or nack).
+   */
+  async stop() {
+    if (!this.running) {
+      return;
+    }
+    this.running = false;
+
+    return this._shutdown();
+  }
+
+  /**
+   * Shut down this listener and channel, without setting this.running to false
+   */
+  async _shutdown() {
+    const {channel, consumerTag} = this;
+
+    if (channel && consumerTag) {
+      this.consumerTag = null;
+      await channel.cancel(consumerTag);
+    }
+
+    // if messages are being processed, arrange to continue when they
+    // are all handled
+    if (this.processingMessages > 0) {
+      await new Promise(resolved => {
+        this.idleCallback = resolved;
+      });
+    }
+
+    if (channel) {
+      this.channel = null;
+      await channel.close();
+    }
+  }
+
+  async _createAndBindQueue(channel) {
+    const queueName = this.client.fullObjectName(
+      'queue',
+      this.queueName || `exclusive/${slugid.v4()}`);
+    await channel.assertQueue(queueName, {
+      exclusive: this.exclusiveQueue,
+      durable: !this.exclusiveQueue,
+      autoDelete: this.exclusiveQueue,
+      maxLength: this.options.maxLength,
+    });
+
+    for (let {exchange, routingKeyPattern} of this.bindings) {
+      await channel.bindQueue(queueName, exchange, routingKeyPattern);
+    }
+
+    return queueName;
+  }
+
+  /**
+   * Handle a new connection to the pulse server, re-declaring everything
+   * and setting up the consumer.
+   */
+  async _handleConnection(conn) {
+    try {
+      if (!this.running) {
+        return;
+      }
+
+      const amqp = conn.amqp;
+      const channel = await amqp.createChannel();
+      await channel.prefetch(this.options.prefetch);
+      const queueName = await await this._createAndBindQueue(channel);
+      this.channel = channel;
+
+      // consider any errors on the channel to be potentially fatal to the
+      // connection (better safe than sorry)
+      channel.on('error', () => conn.failed());
```

No, channel errors are not connection errors.

RabbitMQ is not insane :)

If we have an error at channel, the channel is dead, typically because you made a mistake... But typically, it won't kill the connection, if it does having `.on('error', ` on the connection should be sufficient.

---

I added this at your suggestion, iirc :)

---

So the alternative here may be to repeat the consumer's `createChannel` using the current amqp connection.  In so doing, we need to make sure that we don't set up *two* channels at the same time (once because of a channel error, and the other because of a connection error or periodic recycle).  So we'll need some kind of locking.

I think this is the simpler solution, and given the efforts taken to ensure that a recycle has no application impact, I think it's OK.

Note that you do get an error thrown when you call `channel.someMethod(..)` in an inappropriate way.  So we're not ignoring errors here..

---

differentiating between connection and channel errors by experimenting will make things clear.If channel error is due to some user's invalid data through `channel.someMethod` then repeating `createChannel` may not be right or there should be no. of reattempt limits.